modin/modin_logging/config.py

```diff
@@ -0,0 +1,143 @@
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+import logging
+import datetime as dt
+import os
+import uuid
+import platform
+import psutil
+import pkg_resources
+import threading
+import time
+from modin.config import LogMemoryInterval, LogMode
+
+__LOGGER_CONFIGURED__: bool = False
+
+
+class ModinFormatter(logging.Formatter):
+    def formatTime(self, record, datefmt=None):
+        """
+        Return the creation time of the specified LogRecord as formatted text.
+        This custom logging formatter inherits from the logging module and
+        records timestamps at the microsecond level of granularity.
+
+        Parameters
+        ----------
+        record: LogRecord
+            The specified LogRecord object.
+        datefmt: str, default: None
+            Used with time.ststrftime() to format time record.
+
+        Returns
+        -------
+        datetime
+            datetime object containing microsecond timestamp.
+        """
+        ct = dt.datetime.fromtimestamp(record.created)
+        if datefmt:
+            s = ct.strftime(datefmt)
+        else:
+            # Format datetime object ct to microseconds
+            t = ct.strftime("%Y-%m-%d %H:%M:%S")
+            s = f"{t},{record.msecs:03}"
+        return s
+
+
+def bytes_int_to_str(num_bytes, suffix="B"):
+    """
+    Scale bytes to its human-readable format (e.g: 1253656678 => '1.17GiB').
+
+    Parameters
+    ----------
+    num_bytes: int
+        Number of bytes.
+    suffix: str, default: "B"
+        Suffix to add to conversion of num_bytes.
+
+    Returns
+    -------
+    str
+        Human-readable string format.
+    """
+    factor = 1024
+    if num_bytes > 1000000000000000000:
+        raise ValueError("System memory exceeds expectations")
+    for unit in ["", "Ki", "Mi", "Gi", "Ti", "Pi"]:
+        if num_bytes < factor:
+            return f"{num_bytes:.2f}{unit}{suffix}"
+        num_bytes /= factor
+
+
+def configure_logging(level):
+    global __LOGGER_CONFIGURED__
+    logger = logging.getLogger("modin.logger")
+    job_id = uuid.uuid4().hex
+    log_filename = f".modin/logs/job_{job_id}.log"
+
+    if not os.path.isdir(".modin/logs"):
+        os.makedirs(os.path.dirname(log_filename), exist_ok=False)
+
+    logfile = logging.FileHandler(log_filename, "a")
+    formatter = ModinFormatter(
+        fmt="%(process)d, %(thread)d, %(asctime)s, %(message)s",
+        datefmt="%Y-%m-%d,%H:%M:%S.%f",
+    )
+    logfile.setFormatter(formatter)
+    logger.addHandler(logfile)
+    logger.setLevel(level)
+
+    __LOGGER_CONFIGURED__ = True
+
+
+def memory_thread(logger, sleep_time):
+    while True:
+        svmem = psutil.virtual_memory()
```

Comment:
For my own understanding - if someone has multiple things running on this device, psutil will return memory used by all processes right? Not just the Modin processes? Also, doesn't this only return the memory used on the head node? If the user is running on a cluster this won't really be accurate right? And if they're running on a cluster that doesn't schedule on the head node isn't this going to be super wrong?

Reply:
This is more intended as a general profiling tool (not one specific to Modin, we essentially will use this to identify all memory spikes). Correct, this is going to be useful in most user cases but could be off for some (scheduling on the head node is fairly standard, this will total system memory utilization). To get more specific memory utilization data about the Modin processes, we would have to log inside of the Ray workers which will be done in a future PR by @jeffreykennethli.